CONTRIBUTING.md

```diff
@@ -0,0 +1,261 @@
+
+4. Favor reproducibilty
+
+    Always use fully resolved lockfiles or similar input files to determine what content needs to be download for a specific project (e.g. npm's `package-lock.json`, a `pip-compiled` generated `requirements.txt`, etc). Resolving the dependencies during the prefetch will prevent its behavior from being deterministic, or in other words, the same repository and the same commit hash should always result in identical prefetch results.
+
+## Development
+
+### Virtual environment
+
+Set up a virtual environment that has everything you will need for development:
+
+```shell
+make venv
+source venv/bin/activate
+```
+
+This installs the Cachi2 CLI in [editable mode](https://setuptools.pypa.io/en/latest/userguide/development_mode.html),
+which means changes to the source code will reflect in the behavior of the CLI without the need for reinstalling.
+
+You may need to install Python 3.9 in case you want to test your changes against Python 3.9 locally
+before submitting a pull request.
+
+```shell
+dnf install python3.9
+```
+
+The CLI also depends on the following non-Python dependencies:
+
+```shell
+dnf install golang-bin git
+```
+
+You should now have everything needed to [try out](#basic-usage) the CLI or hack on the code in ~~vim~~ your favorite
+editor.
+
+### Developer flags
+
+* `--dev-package-managers` (hidden): enables in-development package manager(s)
+  for test. Please refer to other existing package managers to see how they're
+  enabled and wired to the CLI.
+
+  Invoke it as `cachi2 fetch-deps --dev-package-managers FOO`
+
+  More explicitly
+
+  * `--dev-package-managers` is a *flag for* `fetch-deps`
+  * `FOO` is an *argument to* `fetch-deps` (i.e. the language to fetch for)
+
+### Coding standards
+
+Cachi2's codebase conforms to standards enforced by a collection of formatters, linters and other code checkers:
+
+* [black](https://black.readthedocs.io/en/stable/) (with a line-length of 100) for consistent formatting
+* [isort](https://pycqa.github.io/isort/) to keep imports sorted
+* [flake8](https://flake8.pycqa.org/en/latest/) to (de-)lint the code and ~~politely~~ ask for docstrings
+* [mypy](https://mypy.readthedocs.io/en/stable/) for type-checking. Please include type annotations for new code.
+* [pytest](https://docs.pytest.org/en/7.1.x/) to run unit tests and report coverage stats. Please aim for (near) full
+  coverage of new code.
+
+Options for all the tools are configured in [pyproject.toml](./pyproject.toml) and [tox.ini](./tox.ini).
+
+Run all the checks that your pull request will be subjected to:
+
+```shell
+make test
+```
+
```

Comment:
Another thing worth mentioning:

"""If your code was inspired by any third-party sources please consider adding a comment with a link to these sources. Generally please consider adding comments to the code whenever there exists any context which is
not obvious from the code alone. When writing a comment do not repeat how a piece of code works, do explain why this is needed."""

Reply:
I don't know where to fit this into. Do you think it's worth adding a small section like "Comment guidelines"?

Reply:
I'd say it is a part of coding standard so it fits right here.

Reply:
Also to correct my own suggestion: s/please consider adding/please add/

Reply:
I moved this to a separate section. My feeling is that it felt out of place in section that is mostly dealing with linters. Since we already had a "error message guidelines" section that felt pretty short, I thought the best solution was to move this piece as a similar section.